DEVINFO.md

```diff
@@ -0,0 +1,174 @@
+
+* Database changes **cannot be consensus-critical** unless part of a hard fork (see below).
+
+* If the database schema changes and no migration can be feasibly done, then the submitter **must spin up a node from genesis to verify that it works** _before_ submitting the PR.  This genesis spin-up will be tested again before the next node release is made.
+
+### Data Input
+
+* **Data from the network, from Bitcoin, and from the config file is untrusted.**  Code that ingests such data _cannot assume anything_ about its structure, and _must_ handle any possible byte sequence that can be submitted to the Stacks node.
+
+* **Data previously written to disk by the node is trusted.** If data loaded from the database that was previously stored by the node is invalid or corrupt, it is appropriate to panic.
+
+* **All input processing is space-bound.**  Every piece of code that ingests data must impose a maximum size on its byte representation.  Any inputs that exceed this size _must be discarded with as little processing as possible_.
+
+* **All input deserialization is resource-bound.** Every piece of code that ingests data must impose a maximum amount of RAM and CPU required to decode it into a structured representation.  If the data does not decode with the allotted resources, then no further processing may be done and the data is discarded.
+
+* **All network input reception is time-bound.**  Every piece of code that ingests data _from the network_ must impose a maximum amount of time that ingestion can take.  If the data takes too long to arrive, then it must be discarded without any further processing.  There is no time bound for data ingested from disk or passed as an argument; this requirement is meant by the space-bound requirement.
+
+* **Untrusted data ingestion must not panic.**  Every piece of code that ingests untrusted data must gracefully handle errors.  Panicking failures are forbidden for such data.  Panics are only allowed if the ingested data was previously written by the node (and thus trusted).
+
+### Non-consensus Changes to Blocks, Microblocks, Transactions, and Clarity
+
+Any changes to code that alters how a block, microblock, or transaction is processed by the node should be **treated as a breaking change until proven otherwise**.  This includes changes to the Clarity VM.  The reviewer _must_ flag any such changes in the PR, and the submitter _must_ convince _all_ reviewers that they will _not_ break consensus.
+
+Changes that touch any of these four code paths must be treated with the utmost care.  If _any_ core developer suspects that a given PR would break consensus, then they _must_ act to prevent the PR from merging.
+
+### Changes to the Peer Network
+
+Any changes to the peer networking code **must be run in production before the PR can be merged.**  The submitter should set up a testable node or set of nodes that reviewers can interact with.
```

Comment:
No, it _must_ run on mainnet before we merge.  If it can't survive running with a production workload in a heterogeneous set of nodes, then it should not be merged.

Reply:
replicable i would say yes, but scalable, i wouldn't say so - I agree with Jude here that the peer networking code should be run on mainnet before the PR can be merged. 
I read this as the PR submitter would follow the expected testing progression of mocknet->testnet->mainnet.

Reply:
I updated this to explicitly say this has to run in both testnet and mainnet.